allennlp/data/dataset_readers/multiprocess_dataset_reader.py

```diff
@@ -0,0 +1,123 @@
+        if file_path is None:
+            break
+
+        logger.info(f"reading instances from {file_path}")
+        for instance in reader.read(file_path):
+            output_queue.put(instance)
+
+    # Put None on the queue to signify that we've finished
+    output_queue.put(None)
+
+
+@DatasetReader.register('multiprocess')
+class MultiprocessDatasetReader(DatasetReader):
+    """
+    Wraps another dataset reader and uses it to read from multiple input files
+    using multiple processes. Note that in this case the ``file_path`` passed to ``read()``
+    should be a glob, and that the dataset reader will return instances from all files
+    matching the glob.
+
+    Parameters
+    ----------
+    base_reader : ``DatasetReader``
+        Each process will use this dataset reader to read zero or more files.
+    num_workers : ``int``
+        How many data-reading processes to run simultaneously.
+    epochs_per_read : ``int``, (optional, default=1)
+        Normally a call to ``DatasetReader.read()`` returns a single epoch worth of instances,
+        and your ``DataIterator`` handles iteration over multiple epochs. However, in the
+        multiple-process case, it's possible that you'd want finished workers to continue on to the
+        next epoch even while others are still finishing the previous epoch. Passing in a value
+        larger than 1 allows that to happen.
+    lazy : ``bool``, (optional, default=True)
+        Most of our dataset readers are eager by default; however, if you're using this one
+        it's probably because you have a lot of data (and in particular don't want to load it
+        all into memory), so laziness is the default.
+    """
+    def __init__(self,
+                 base_reader: DatasetReader,
+                 num_workers: int,
+                 epochs_per_read: int = 1,
+                 lazy: bool = True) -> None:
+        super().__init__(lazy=lazy)
+
+        self.reader = base_reader
+        self.num_workers = num_workers
+        self.epochs_per_read = epochs_per_read
+
+    def text_to_instance(self, *args, **kwargs) -> Instance:
+        """
+        Just delegate to the base reader text_to_instance.
+        """
+        # pylint: disable=arguments-differ
+        return self.reader.text_to_instance(*args, **kwargs)
+
+    def _read(self, file_path: str) -> Iterable[Instance]:
+        shards = glob.glob(file_path)
+        num_shards = len(shards)
+
+        # If we want multiple epochs per read, put shards in the queue multiple times.
+        input_queue = Queue(num_shards * self.epochs_per_read + self.num_workers)
+        for _ in range(self.epochs_per_read):
+            random.shuffle(shards)
+            for shard in shards:
+                input_queue.put(shard)
+
+        # Then put a None per worker to signify no more files.
+        for _ in range(self.num_workers):
+            input_queue.put(None)
+
+        # TODO(joelgrus): where does this number come from?
```

Comment:
We could allow this to be configurable, and may need to increase it depending on how many instances the iterator uses for each batch.  We want it to be large enough allow a significant buffer so the iterator doesn't have to wait for instances but not too large that it starts taking up significant memory.